File: ALQASEER-PWA/CRM/backend/api/v1/collections.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from api.v1.utils import DEFAULT_PAGE, DEFAULT_PAGE_SIZE, clamp_page_size, paginate
from core.db import get_db
from core.security import get_current_user, require_roles
from models.crm import Collection, Doctor, Pharmacy
from schemas.common import PaginatedResponse
from schemas.crm import CollectionCreate, CollectionOut

router = APIRouter(
    prefix="/collections",
    tags=["collections"],
    dependencies=[Depends(get_current_user)],
)


@router.get("/", response_model=PaginatedResponse[CollectionOut])
def list_collections(
    page: int = Query(DEFAULT_PAGE, ge=1),
    page_size: int = Query(DEFAULT_PAGE_SIZE, ge=1, le=500),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
) -> PaginatedResponse[CollectionOut]:
    query = db.query(Collection)
    if date_from:
        query = query.filter(Collection.collection_date >= date_from)
    if date_to:
        query = query.filter(Collection.collection_date <= date_to)

    page_size = clamp_page_size(page_size)
    items, total = paginate(query.order_by(Collection.collection_date.desc()), page, page_size)
    total_pages = max(1, (total + page_size - 1) // page_size)
    return PaginatedResponse(
        data=items,
        pagination={"page": page, "page_size": page_size, "total": total, "total_pages": total_pages},
    )


@router.post(
    "/",
    response_model=CollectionOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles("sales_manager", "admin", "accountant"))],
)
def create_collection(payload: CollectionCreate, db: Session = Depends(get_db)) -> Collection:
    if payload.doctor_id and not db.get(Doctor, payload.doctor_id):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Doctor not found.")
    if payload.pharmacy_id and not db.get(Pharmacy, payload.pharmacy_id):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Pharmacy not found.")
    collection = Collection(**payload.model_dump())
    db.add(collection)
    db.commit()
    db.refresh(collection)
    return collection

File: CRM/frontend/src/api/collections.js
import { apiClient } from './client';
import { normalizePaginatedResponse } from './pagination';

export const collectionKeys = {
  all: ['collections'],
  list: filters => ['collections', filters],
};

export const listCollections = async (filters = {}) => {
  const params = new URLSearchParams();
  if (filters.page) params.set('page', String(filters.page));
  if (filters.page_size) params.set('page_size', String(filters.page_size));
  if (filters.date_from) params.set('date_from', filters.date_from);
  if (filters.date_to) params.set('date_to', filters.date_to);
  const qs = params.toString();
  const { data } = await apiClient.get(qs ? `/collections/?${qs}` : '/collections/');
  return normalizePaginatedResponse(data);
};

export const createCollection = async payload => {
  const { data } = await apiClient.post('/collections/', { body: payload });
  return data?.data || data;
};
